example/heatEquation2D/src/BoundaryKernel.hpp

```diff
@@ -0,0 +1,98 @@
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const& acc,
+        double* const uBuf,
+        TChunk const chunkSize,
+        TChunk const pitch,
+        uint32_t step,
+        double const dx,
+        double const dy,
+        double const dt) const -> void
+    {
+        using Dim = alpaka::DimInt<2u>;
+        using Idx = uint32_t;
+
+        // Get extents(dimensions)
+        auto const gridBlockExtent = alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc);
+        auto const blockThreadExtent = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc);
+        auto const numThreadsPerBlock = blockThreadExtent.prod();
+
+        // Get indexes
+        auto const gridBlockIdx = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc);
+        auto const blockThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc);
+        auto const threadIdx1D = alpaka::mapIdx<1>(blockThreadIdx, blockThreadExtent)[0u];
+        auto const blockStartIdx = gridBlockIdx * chunkSize;
+
+        // apply boundary conditions
+        // top row
+        if(gridBlockIdx[0] == 0)
+        {
+            auto const globalIdx = blockStartIdx + alpaka::Vec<Dim, Idx>(0, 1);
+            for(auto i = threadIdx1D; i < chunkSize[1]; i += numThreadsPerBlock)
+            {
+                auto idx2D = globalIdx + alpaka::Vec<Dim, Idx>(0, i);
+                auto elem = getElementPtr(uBuf, idx2D, pitch);
+                *elem = exactSolution(idx2D[1] * dx, idx2D[0] * dy, step * dt);
+            }
+        }
+        if(gridBlockIdx[0] == gridBlockExtent[0] - 1)
+        {
+            auto const globalIdx = blockStartIdx + alpaka::Vec<Dim, Idx>{chunkSize[0] + 1, 1};
+
+            for(auto i = threadIdx1D; i < chunkSize[1]; i += numThreadsPerBlock)
+            {
+                auto idx2D = globalIdx + alpaka::Vec<Dim, Idx>(0, i);
+                auto elem = getElementPtr(uBuf, idx2D, pitch);
+                *elem = exactSolution(idx2D[1] * dx, idx2D[0] * dy, step * dt);
+            }
+        }
+        if(gridBlockIdx[1] == 0)
+        {
+            auto const globalIdx = blockStartIdx + alpaka::Vec<Dim, Idx>(1, 0);
+
+            for(auto i = threadIdx1D; i < chunkSize[0]; i += numThreadsPerBlock)
+            {
+                auto idx2D = globalIdx + alpaka::Vec<Dim, Idx>(i, 0);
+                auto elem = getElementPtr(uBuf, idx2D, pitch);
+                *elem = exactSolution(idx2D[1] * dx, idx2D[0] * dy, step * dt);
+            }
+        }
+        if(gridBlockIdx[1] == gridBlockExtent[1] - 1)
+        {
+            auto const globalIdx = blockStartIdx + alpaka::Vec<Dim, Idx>{1, chunkSize[1] + 1};
+
+            for(auto i = threadIdx1D; i < chunkSize[0]; i += numThreadsPerBlock)
+            {
+                auto idx2D = globalIdx + alpaka::Vec<Dim, Idx>(i, 0);
+                auto elem = getElementPtr(uBuf, idx2D, pitch);
+                *elem = exactSolution(idx2D[1] * dx, idx2D[0] * dy, step * dt);
+            }
+        }
```

Comment:
```suggestion
             // Lambda function to apply boundary conditions
        auto applyBoundary = [&](auto const& globalIdxStart, auto const length, bool isRow) {
            for (auto i = threadIdx1D; i < length; i += numThreadsPerBlock)
            {
                alpaka::Vec<Dim, Idx> idx2D = globalIdxStart + (isRow ? alpaka::Vec<Dim, Idx>{0, i} : alpaka::Vec<Dim, Idx>{i, 0});
                auto elem = getElementPtr(uBuf, idx2D, pitch);
                *elem = exactSolution(idx2D[1] * dx, idx2D[0] * dy, step * dt);
            }
        };

               // Apply boundary conditions for the top row
        if (gridBlockIdx[0] == 0) {
            applyBoundary(blockStartIdx + alpaka::Vec<Dim, Idx>{0, 1}, chunkSize[1], true);
        }

               // Apply boundary conditions for the bottom row
        if (gridBlockIdx[0] == gridBlockExtent[0] - 1) {
            applyBoundary(blockStartIdx + alpaka::Vec<Dim, Idx>{chunkSize[0] + 1, 1}, chunkSize[1], true);
        }

               // Apply boundary conditions for the left column
        if (gridBlockIdx[1] == 0) {
            applyBoundary(blockStartIdx + alpaka::Vec<Dim, Idx>{1, 0}, chunkSize[0], false);
        }

               // Apply boundary conditions for the right column
        if (gridBlockIdx[1] == gridBlockExtent[1] - 1) {
            applyBoundary(blockStartIdx + alpaka::Vec<Dim, Idx>{1, chunkSize[1] + 1}, chunkSize[0], false);
        }
```

Reply:
@ikbuibui Could you take this suggestion in and rebase and squash your code. 32 commits is a lot for this example.

Reply:
done